lib/src/str_util.rs

```diff
@@ -32,105 +32,189 @@ pub enum StringPatternParseError {
-    /// `StringPattern::Exact` is returned.
+    /// Helper for `StringPattern::parse_with_default_kind(src, "exact')`. See
+    /// [`StringPattern::parse_with_default_kind()`] for details.
     pub fn parse(src: &str) -> Result<StringPattern, StringPatternParseError> {
-        if let Some((kind, pat)) = src.split_once(':') {
-            StringPattern::from_str_kind(pat, kind)
-        } else {
-            Ok(StringPattern::exact(src))
-        }
+        Self::parse_with_default_kind(src, "exact")
     }
 
     /// Creates pattern that matches exactly.
-    pub fn exact(src: impl Into<String>) -> Self {
-        StringPattern::Exact(src.into())
+    pub fn exact(src: impl Into<String>, case_sensitivity: CaseSensitivity) -> Self {
+        StringPattern::Exact(src.into(), case_sensitivity)
     }
 
     /// Parses the given string as glob pattern.
-    pub fn glob(src: &str) -> Result<Self, StringPatternParseError> {
+    pub fn glob(
+        src: &str,
+        case_sensitivity: CaseSensitivity,
+    ) -> Result<Self, StringPatternParseError> {
         // TODO: might be better to do parsing and compilation separately since
         // not all backends would use the compiled pattern object.
         // TODO: if no meta character found, it can be mapped to Exact.
         let pattern = glob::Pattern::new(src).map_err(StringPatternParseError::GlobPattern)?;
-        Ok(StringPattern::Glob(pattern))
+        Ok(StringPattern::Glob(pattern, case_sensitivity))
     }
 
-    /// Parses the given string as pattern of the specified `kind`.
+    /// Parses the given string as a pattern of the specified `kind`.
     pub fn from_str_kind(src: &str, kind: &str) -> Result<Self, StringPatternParseError> {
-        match kind {
-            "exact" => Ok(StringPattern::exact(src)),
-            "glob" => StringPattern::glob(src),
-            "substring" => Ok(StringPattern::Substring(src.to_owned())),
+        let (base_kind, case_sensitivity) = match kind.strip_suffix("-i") {
+            Some(base_kind) => (base_kind, CaseSensitivity::Insensitive),
+            None => (kind, CaseSensitivity::Sensitive),
+        };
+        match base_kind {
+            "exact" => Ok(StringPattern::exact(src, case_sensitivity)),
+            "substring" => Ok(StringPattern::Substring(src.to_owned(), case_sensitivity)),
+            "glob" => StringPattern::glob(src, case_sensitivity),
             _ => Err(StringPatternParseError::InvalidKind(kind.to_owned())),
         }
     }
 
+    /// Parses the given string as a pattern of the specified `maybe_kind` if
+    /// present, or `default_kind` otherwise.
+    pub fn from_str_maybe_kind(
+        src: &str,
+        maybe_kind: Option<&str>,
+        default_kind: &str,
+    ) -> Result<Self, StringPatternParseError> {
+        match maybe_kind {
+            None => Self::from_str_kind(src, default_kind),
+            Some("i") => Self::from_str_kind(src, &(default_kind.to_owned() + "-i")),
```

Comment:
I think it's less confusing if "i:" is always mapped to "contains-i:".

It's unfortunate that we have two defaults, "substring:"/"exact:". It'll be nice if we can eventually pick "exact:" as the default, but I'm not sure if that level of breaking change is acceptable.

Reply:
Hmm, it’d be pretty weird if putting `i:` in front of a branch pattern suddenly started operating on other branches because it turned into a substring match, wouldn’t it? On the other hand, for log queries, having to write `description(substring:todo)` (or even `contains:todo`) and not being able to do `author(martinvonz)` for log queries sounds really inconvenient (though I guess we could make aliases like `c:`, `ci:`, but it seems a bit obscure).

I do really dislike the inconsistency, but I’m not sure what the right thing would be here, even long‐term. Here’s my best idea: maybe glob should be the default, so that matching is exact by default, substring matching is as easy as `*foo*` for simple queries, and you only have to worry about the syntax if your query contains one of `?*[`? Then `i:` could be `glob-i:`, of course. Conveniently, globs are the most advanced syntax you can use for Git branches (because of protocol restrictions).

Reply:
I forgot to mention that moving to that default would of course have to include optimizing globs to exact patterns where possible, but that should be easy given I already implemented an equivalent check without issues (even just using `Exact` when there are no meta‐characters in the string should suffice for ≥90% of cases).

Reply:
I couldn’t resist and tried implementing the `glob` default locally. I like it! I think it’s probably the most viable universal default we could move to, and I’d be happy to finish off the work. It’s clearly too big and too breaking a change to do here, though, so I’d like to propose keeping the current PR behaviour for now. It’s the least surprising and most compatible thing given the existing inconsistency, and we can revisit the question afterwards, since any resolution to the inconsistency would break much more anyway.

Reply:
I assumed that if pattern kind is specified in e.g. `jj branch delete i:foo`, the user would want to do some non-exact matching. And it would be easier to understand if `i:` is documented as short for `substring-i:`, not an context-dependent prefix. (If the syntax were `/i:`, it would look like short for `<default>/i:`, though.)

Maybe better to not add `i:` because at least two folks recognize it differently?

> Here’s my best idea: maybe glob should be the default, so that matching is exact by default,

It might be okay in revset, but I wouldn't expect CLI arguments are glob by default. It would be weird that `jj branch create NAME` isn't inherently glob, but `jj branch delete NAME` were.

Reply:
I’ve removed `i:`; the ergonomics aren’t fantastic now, but the foundation is in place for us to decide what we want the syntax/semantics to look like later, and I mostly just wanted this plumbing internally for the `.mailmap` work anyway.

I personally don’t find `create` and `delete` taking different syntaxes that weird; the ambiguity already exists for `jj branch {create,delete} exact:weird-name`, and glob syntax in branch names is pretty marginal (not sure how it would even works with refspecs?). Creation and queries usually use different syntax, after all. But hopefully we can bikeshed what the syntax and semantics should be separately to this PR :) The changes should be simpler now; I’d appreciate it if you could take another look!